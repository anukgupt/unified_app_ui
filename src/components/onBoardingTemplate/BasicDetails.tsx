import * as React from 'react';
import '../../css/onBoarding/Template.css';
import { onBoardingObject } from "./OnBoardingTemplate";
import * as Constants from "../../Constants";

interface IBasicDetailsProps {
    onBoardingTemplateObject: onBoardingObject;
    onClientIdChange: any;
    onWebhookURLChange: any;
    onCallbackURLChange: any;
}

class BasicDetails extends React.Component<IBasicDetailsProps> {
    render() {
        var clientIdClassName = "client-id";
        if (this.props.onBoardingTemplateObject.clientIdError) {
            var clientIdClassName = "client-id margin-bottom-on-error";
        }
        return (
            <div className="basic-details">
                <div className={clientIdClassName}>
                    <div className="">
                        <div className="subHead">{Constants.ClientId}</div>
                    </div>
                    <dl>
                        <dt>
                            <label htmlFor="clientId" className="input-label">{Constants.ClientId}</label>
                        </dt>
                        <dd>
                            <input className="input-text" type="text" id="clientId" value={this.props.onBoardingTemplateObject.clientId} onChange={this.props.onClientIdChange}></input>
                            <p className="note">{Constants.ClientIdDescription}</p>
                        </dd>
                        {
                            this.props.onBoardingTemplateObject.clientIdError &&
                            <dd className="error">
                                {Constants.ClientIdError}
                            </dd>
                        }
                    </dl>
                </div>
                <div className="webhook-url">
                    <div className="">
                        <div className="subHead">{Constants.Webhook}</div>
                    </div>
                    <dl>
                        <dt>
                            <label htmlFor="webhookURL" className="input-label">{Constants.WebhookURL}</label>
                        </dt>
                        <dd>
                            <input className="input-text" type="text" id="webhookURL" value={this.props.onBoardingTemplateObject.webhookURL} onChange={this.props.onWebhookURLChange}></input>
                            <p className="note">
                                {Constants.WebhookURLDescription1}
                                <a target="_blank" href={Constants.WebhookURLLink}>{Constants.WebhookURLDescriptionLink}</a>
                                {Constants.WebhookURLDescription2}
                            </p>
                        </dd>
                    </dl>
                </div>
                <div className="callback-url">
                    <div className="">
                        <div className="subHead">{Constants.Callback}</div>
                    </div>
                    <dl>
                        <dt>
                            <label htmlFor="callbackURL" className="input-label">{Constants.CallbackURL}</label>
                        </dt>
                        <dd>
                            <input className="input-text" type="text" id="callbackURL" value={this.props.onBoardingTemplateObject.callbackURL} onChange={this.props.onCallbackURLChange}></input>
                            <p className="note">{Constants.CallbackURLDescription}</p>
                        </dd>
                    </dl>
                </div>
            </div>
        )
    }
}
export default BasicDetails;