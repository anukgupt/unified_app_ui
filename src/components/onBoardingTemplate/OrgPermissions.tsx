import * as React from 'react';
import '../../css/onBoarding/Template.css';
import { onBoardingObject } from "./OnBoardingTemplate";
import icon from '../../images/icon.png';
import down from '../../images/downarrow.png';
import cross from "../../images/cross.png";
import tick from "../../images/tick.png";
import { SelectMenu, Button } from '@primer/components';
import * as Constants from "../../Constants";

const organizationPermissions = [
    { "id": "members", "text": "Members", "note": "Organization members and teams.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-members" },
    { "id": "organizationadministration", "text": "Administration", "note": "Manage access to an organization.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-organization-administration" },
    { "id": "organizationwebhooks", "text": "Webhooks", "note": "Manage the post-receive hooks for an organization.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-organization-hooks" },
    { "id": "organizationplan", "text": "Plan", "note": "View an organization's plan.", "accessLevel": ["No access", "Read-only"], "link": "https://docs.github.com/v3/apps/permissions/" },
    { "id": "organizationprojects", "text": "Projects", "note": "Manage organization projects, columns, and cards.", "accessLevel": ["No access", "Read-only", "Read & Write", "Admin"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-organization-projects" },
    { "id": "organizationsecrets", "text": "Secrets", "note": "Manage organization secrets.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/" },
    { "id": "selfhostedrunners", "text": "Self-hosted runners", "note": "View and manage Actions self-hosted runners available to an organization.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/" },
    { "id": "blockingusers", "text": "Blocking users", "note": "View and manage users blocked by the organization.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-organization-user-blocking" },
    { "id": "teamdiscussions", "text": "Team discussions", "note": "Manage team discussions and related comments.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-team-discussions" }
];
interface IOrgPermissionsProps {
    setTemplateState: any;
    onBoardingTemplateObject: onBoardingObject;
}
class OrgPermissions extends React.Component<IOrgPermissionsProps> {
    constructor(props: any) {
        super(props);
        this.initializeOrgPermissions();
    }

    initializeOrgPermissions() {
        let onBoardingTemplateObject = this.props.onBoardingTemplateObject;
        organizationPermissions.map(organizationPermission => {
            onBoardingTemplateObject.orgPermissions[organizationPermission.id] = organizationPermission.accessLevel[0]
        })
        this.props.setTemplateState(onBoardingTemplateObject);
    }

    render() {
        let orgPerm: any = [];
        let accessLevel: any = [];
        organizationPermissions.map(organizationPermission => {
            accessLevel = [];
            organizationPermission.accessLevel.map((access, index) => {
                let hidden = true;
                if (access === this.props.onBoardingTemplateObject.orgPermissions[organizationPermission.id]) {
                    hidden = false;
                }
                accessLevel.push(
                    <SelectMenu.Item style={{ 'height': '24px', 'padding': '0px' }} key={organizationPermission.text + "" + access} id={organizationPermission.id + index} onClick={() => this.onAccessLevelClick(organizationPermission.id, access)}>
                        <div className="input-access">
                            <div className="input-access-tick">
                                <img aria-label="checked" hidden={hidden} className="tick-img" alt="" src={tick} width="16px" height="16px" />
                            </div>
                            <div className="input-access-text">
                                {access}
                            </div>
                        </div>
                    </SelectMenu.Item >
                );
            });
            orgPerm.push(
                <div key={organizationPermission.id} className="org-permission-item tableobject">
                    <div className="tableobject-cell-primary">
                        <div>{organizationPermission.text}
                            <a target="_blank" className="link" aria-label="info" href={organizationPermission.link}>
                                <img className="info-img" alt="" src={icon} width="16px" height="16px" />
                            </a>
                        </div>
                        <p className="note">{organizationPermission.note}</p>
                    </div>
                    <div className="tableobject-cell-intermediate">
                        Access:
                    </div>
                    <div className="tableobject-cell-secondary">
                        <SelectMenu className="select-menu">
                            <Button as="summary" className="selectmenu-button" style={{ 'padding': '2px', 'width': '180px', 'border': ' 1px solid #8A8886', 'borderRadius': '2px', 'backgroundColor': '#fff', 'lineHeight': '18px' }}>
                                <span className="selectmenu-button-text">{this.props.onBoardingTemplateObject.orgPermissions[organizationPermission.id]}</span>
                                <img className="down-img" alt="" src={down} />
                            </Button>
                            <SelectMenu.Modal width="185px">
                                <SelectMenu.Header style={{ 'padding': '0', 'height': '24px' }}>
                                    <span className="select-menu-header-text">{Constants.SelectMenuHeading}</span>
                                    <this.AccessLevelCloseButton />
                                </SelectMenu.Header>
                                <SelectMenu.List>
                                    {accessLevel}
                                </SelectMenu.List>
                            </SelectMenu.Modal>
                        </SelectMenu>
                    </div>
                </div>
            )
        });
        return (
            <div className="permissions">
                <div className="organization-permissions" aria-label="organization-permissions">
                    {orgPerm}
                </div>
            </div >

        )
    }

    onAccessLevelClick(id: string, accessValue: string) {
        let onBoardingTemplateObject = this.props.onBoardingTemplateObject;
        onBoardingTemplateObject.orgPermissions[id] = accessValue;
        this.props.setTemplateState(onBoardingTemplateObject);
    }

    AccessLevelCloseButton = () => {
        const menuContext = React.useContext(SelectMenu.MenuContext);
        return (
            <button type="button" aria-label="close menu" className="cross-img-button" onClick={() => {
                menuContext.setOpen(false);
            }}>
                <img className="cross-img" alt="" src={cross} width="10px" height="10px" />
            </button>
        );
    }
}
export default OrgPermissions