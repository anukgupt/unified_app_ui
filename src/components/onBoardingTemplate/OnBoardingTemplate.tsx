import * as React from 'react';
import '../../css/onBoarding/Template.css';
import * as Constants from "../../Constants";
import withAuthProvider from "../../provider/AuthProviderForOnBoarding";
import BasicDetails from './BasicDetails';
import RepoPermissions from './RepoPermissions';
import OrgPermissions from './OrgPermissions';
import UserPermissions from './UserPermissions';
import { saveTemplateMapping } from '../../services/onBoardingTemplate/TemplateMappingService';
import one from "../../images/one.png";
import two from "../../images/two.png";
import three from "../../images/three.png";
import four from "../../images/four.png";
import error from "../../images/error.png";
import successRest from "../../images/successRest.png";
import lessthan from "../../images/lessthan.png";
import greaterthan from "../../images/greaterthan.png";
import { Tabs, TabPanel, Tab, TabList } from 'react-tabs';

export interface onBoardingObject {
    clientId: string;
    clientIdError: boolean;
    webhookURL: string;
    webhookURLError: boolean;
    callbackURL: string;
    callbackURLError: boolean;
    repoPermissions: { [key: string]: string };
    orgPermissions: { [key: string]: string };
    userPermissions: { [key: string]: string };
    showOnBoardingTemplate: boolean;
    isUserAuthorized: boolean;
    isTemplateSucessfullySaved: boolean;
    selectedTabIndex: number;
    isUserAdminFlowInProcess: boolean;
}
interface ITemplateState {
    onBoardingTemplateObject: onBoardingObject
}

class OnBoardingTemplate extends React.Component<any, ITemplateState> {
    private accessToken: string = '';
    constructor(props: any) {
        super(props);
        let onBoardingTemplateObject: onBoardingObject = {
            clientId: "",
            clientIdError: false,
            webhookURL: "",
            webhookURLError: false,
            callbackURL: "",
            callbackURLError: false,
            repoPermissions: {},
            orgPermissions: {},
            userPermissions: {},
            showOnBoardingTemplate: true,
            isUserAuthorized: false,
            isTemplateSucessfullySaved: false,
            selectedTabIndex: 0,
            isUserAdminFlowInProcess: true
        };
        this.state = {
            onBoardingTemplateObject: onBoardingTemplateObject
        };
        this.initialize();
    }

    async initialize() {
        let onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        try {
            // this.accessToken = await this.props.login('');
            // onBoardingTemplateObject.isUserAuthorized = this.props.isAuthenticated;
            onBoardingTemplateObject.isUserAuthorized = true;
            onBoardingTemplateObject.isUserAdminFlowInProcess = false;
            this.setTemplateState(onBoardingTemplateObject);
        }
        catch (error) {
            onBoardingTemplateObject.isUserAdminFlowInProcess = false;
            console.log(error);
        }
    }

    render(): React.ReactNode {
        let buttonClassName = "buttons"
        if (this.state.onBoardingTemplateObject.isUserAuthorized && this.state.onBoardingTemplateObject.selectedTabIndex == -1) {
            buttonClassName = "buttons buttons-position";
        }
        let oneIcon = one, twoIcon = two, threeIcon = three, fourIcon = four;
        if (this.state.onBoardingTemplateObject.isUserAuthorized && this.state.onBoardingTemplateObject.isTemplateSucessfullySaved) {
            oneIcon = twoIcon = threeIcon = fourIcon = successRest;
        }
        let tabs: any[] = [];
        tabs[0] = tabs[1] = tabs[2] = tabs[3] = "tab";
        tabs[this.state.onBoardingTemplateObject.selectedTabIndex] = "tab underline";
        return (
            !this.state.onBoardingTemplateObject.isUserAdminFlowInProcess &&
            <div className="main">
                {
                    <div className="container-1">
                        <div className="onboarding-template">
                            <div className="microsoft-azure">
                                <span className="microsoft-azure-heading">{Constants.MicrosoftAzureHeading}</span>
                            </div>
                            <div className="onboarding-heading">
                                <div className="onboarding-heading-text">{Constants.OnBoardingHeading}</div>
                                <div className="onboarding-heading-create">{Constants.CreateHeading}</div>
                            </div>
                            {
                                this.state.onBoardingTemplateObject.isUserAuthorized &&
                                <form id="onboarding-template" action="">
                                    <div className="onboarding-form">
                                        <Tabs selectedIndex={this.state.onBoardingTemplateObject.selectedTabIndex} onSelect={() => { }}>
                                            <TabList className="tab-list">
                                                <Tab className={tabs[0]} href="" aria-label={Constants.ConfigurationHeading} onClick={() => this.onSelectingTab(0)}>
                                                    <img className="tab-image" alt="" src={oneIcon} />
                                                    <span className="tab-text">{Constants.ConfigurationHeading}</span>
                                                </Tab>
                                                <Tab className={tabs[1]} href="" aria-label={Constants.RepoPermissionsHeading} onClick={() => this.onSelectingTab(1)}>
                                                    <img className="tab-image" alt="" src={twoIcon} />
                                                    <span> {Constants.RepoPermissionsHeading}</span>
                                                </Tab>
                                                <Tab className={tabs[2]} href="" aria-label={Constants.OrgPermissionsHeading} onClick={() => this.onSelectingTab(2)}>
                                                    <img className="tab-image" alt="" src={threeIcon} />
                                                    <span> {Constants.OrgPermissionsHeading}</span>
                                                </Tab>
                                                <Tab className={tabs[3]} href="" aria-label={Constants.UserPermissionsHeading} onClick={() => this.onSelectingTab(3)}>
                                                    <img className="tab-image" alt="" src={fourIcon} />
                                                    <span> {Constants.UserPermissionsHeading}</span>
                                                </Tab>
                                            </TabList>
                                            <TabPanel><BasicDetails onBoardingTemplateObject={this.state.onBoardingTemplateObject} onClientIdChange={this.onClientIdChange} onWebhookURLChange={this.onWebhookURLChange} onCallbackURLChange={this.onCallbackURLChange} /></TabPanel>
                                            <TabPanel><RepoPermissions setTemplateState={this.setTemplateState} onBoardingTemplateObject={this.state.onBoardingTemplateObject} /></TabPanel>
                                            <TabPanel><OrgPermissions setTemplateState={this.setTemplateState} onBoardingTemplateObject={this.state.onBoardingTemplateObject} /></TabPanel>
                                            <TabPanel><UserPermissions setTemplateState={this.setTemplateState} onBoardingTemplateObject={this.state.onBoardingTemplateObject} /></TabPanel>
                                        </Tabs>
                                        {
                                            this.state.onBoardingTemplateObject.isUserAuthorized && this.state.onBoardingTemplateObject.isTemplateSucessfullySaved && this.state.onBoardingTemplateObject.selectedTabIndex == -1 &&
                                            < div className="template-saved">
                                                <img className="template-saved-image" alt="" src={successRest} />
                                                <div className="template-saved-message">
                                                    <span>{Constants.TemplateSuccessfullySavedMessage}</span>
                                                    <span><a className="template-saved-link" href={Constants.TemplateSuccessfullySavedLink}>{Constants.TemplateSuccessfullySavedLinkText}</a></span>
                                                    <span>.</span>
                                                </div>
                                            </div>
                                        }
                                        {
                                            this.state.onBoardingTemplateObject.isUserAuthorized && !this.state.onBoardingTemplateObject.isTemplateSucessfullySaved && this.state.onBoardingTemplateObject.selectedTabIndex == -1 &&
                                            < div className="template-not-saved">
                                                <img className="template-not-saved-image" alt="" src={error} />
                                                <div className="template-not-saved-message">
                                                    <span>{Constants.TemplateNotSavedMessage}</span>
                                                </div>
                                            </div>
                                        }
                                    </div>
                                    <div className={buttonClassName}>
                                        <div className="all-buttons">
                                            <button type="button" className="create-configuration" aria-label={Constants.CreateOnboardingTemplateButton} onClick={this.submitTemplate}>
                                                <span className="create-configuration-text">{Constants.CreateOnboardingTemplateButton}</span>
                                            </button>
                                            <button type="button" className="previous-button" aria-label={Constants.PreviousButton} onClick={this.onPreviousButtonClick}>
                                                <img className="button-image-prev" alt="" src={lessthan} />
                                                <span>{Constants.PreviousButton}</span>
                                            </button>
                                            <button type="button" className="next-button" aria-label={Constants.NextButton} onClick={this.onNextButtonClick}>
                                                <span>{Constants.NextButton}</span>
                                                <img className="button-image-next" alt="" src={greaterthan} />
                                            </button>
                                        </div>
                                    </div>
                                </form>
                            }
                            {
                                !this.state.onBoardingTemplateObject.isUserAuthorized &&
                                < div className="user-not-authorized">
                                    <img className="user-not-authorized-image" alt="" src={error} />
                                    <div className="user-not-authorized-message">
                                        <span>{Constants.UserNotAuthorizedMessage}</span>
                                        <span><a className="user-not-authorized-link" href={Constants.UserNotAuthorizedLink}>{Constants.UserNotAuthorizedLinkText}</a></span>
                                        <span>.</span>
                                    </div>
                                </div>
                            }
                        </div>
                    </div >
                }
            </div >
        )
    }

    setTemplateState = (onBoardingTemplateObject: onBoardingObject) => {
        this.setState({ onBoardingTemplateObject: onBoardingTemplateObject });
    }

    onClientIdChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        let onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        onBoardingTemplateObject.clientId = event.target.value;
        this.setTemplateState(onBoardingTemplateObject);
    }

    onWebhookURLChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        let onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        onBoardingTemplateObject.webhookURL = event.target.value;
        this.setTemplateState(onBoardingTemplateObject);
    }

    onCallbackURLChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        let onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        onBoardingTemplateObject.callbackURL = event.target.value;
        this.setTemplateState(onBoardingTemplateObject);
    }

    onRepPermissionsChange = (id: string, value: string) => {
        let onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        onBoardingTemplateObject.repoPermissions[id] = value;
        this.setTemplateState(onBoardingTemplateObject);
    }

    onOrgPermissionsChange = (id: string, value: string) => {
        let onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        onBoardingTemplateObject.orgPermissions[id] = value;
        this.setTemplateState(onBoardingTemplateObject);
    }

    onUserPermissionsChange = (id: string, value: string) => {
        let onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        onBoardingTemplateObject.userPermissions[id] = value;
        this.setTemplateState(onBoardingTemplateObject);
    }

    validationsOnTemplateSubmission(): onBoardingObject {
        let onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        if (this.state.onBoardingTemplateObject.clientId.length === 0) {
            onBoardingTemplateObject.clientIdError = true;
        }
        else
            onBoardingTemplateObject.clientIdError = false;
        return onBoardingTemplateObject;
    }

    submitTemplate = () => {
        var onBoardingTemplateObject = this.validationsOnTemplateSubmission();
        if (onBoardingTemplateObject.clientIdError) {
            this.setTemplateState(onBoardingTemplateObject);
            document.getElementById("clientId")?.focus();
        }
        else
            try {
                saveTemplateMapping(this.state.onBoardingTemplateObject, this.accessToken);
                onBoardingTemplateObject.showOnBoardingTemplate = false;
                onBoardingTemplateObject.isTemplateSucessfullySaved = true;
                onBoardingTemplateObject.selectedTabIndex = -1;
                this.setTemplateState(onBoardingTemplateObject);
            }
            catch (error) {
                onBoardingTemplateObject.isTemplateSucessfullySaved = false;
                onBoardingTemplateObject.showOnBoardingTemplate = false;
                onBoardingTemplateObject.selectedTabIndex = -1;
                this.setTemplateState(onBoardingTemplateObject);
            }
    }

    onNextButtonClick = () => {
        var onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        onBoardingTemplateObject.selectedTabIndex = onBoardingTemplateObject.selectedTabIndex + 1;
        if (onBoardingTemplateObject.selectedTabIndex > 3) {
            onBoardingTemplateObject.selectedTabIndex = 3;
        }
        this.setTemplateState(onBoardingTemplateObject);
    }

    onPreviousButtonClick = () => {
        var onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        onBoardingTemplateObject.selectedTabIndex = onBoardingTemplateObject.selectedTabIndex - 1;
        if (onBoardingTemplateObject.selectedTabIndex < 0) {
            onBoardingTemplateObject.selectedTabIndex = 0;
        }
        this.setTemplateState(onBoardingTemplateObject);
    }

    onSelectingTab(index: number) {
        var onBoardingTemplateObject = this.state.onBoardingTemplateObject;
        onBoardingTemplateObject.selectedTabIndex = index;
        this.setTemplateState(onBoardingTemplateObject);
    }
}

export default withAuthProvider(OnBoardingTemplate);