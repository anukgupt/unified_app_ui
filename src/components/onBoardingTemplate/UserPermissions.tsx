import * as React from 'react';
import '../../css/onBoarding/Template.css';
import { onBoardingObject } from "./OnBoardingTemplate";
import icon from '../../images/icon.png';
import down from '../../images/downarrow.png';
import cross from "../../images/cross.png";
import tick from "../../images/tick.png";
import { SelectMenu, Button } from '@primer/components';
import * as Constants from "../../Constants";

const userPermissions = [
    { "id": "blockanotheruser", "text": "Block another user", "note": "View and manage users blocked by the user.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-blocking" },
    { "id": "emailaddresses", "text": "Email addresses", "note": "Manage a user's email addresses.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-emails" },
    { "id": "followers", "text": "Followers", "note": "A user's followers", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-followers" },
    { "id": "gpgkeys", "text": "GPG keys", "note": "View and manage a user's GPG keys.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-gpg-keys" },
    { "id": "gitsshkeys", "text": "Git SSH keys", "note": "Git SSH keys", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-keys" },
    { "id": "plan", "text": "Plan", "note": "View a user's plan.", "accessLevel": ["No access", "Read-only"], "link": "https://docs.github.com/v3/apps/permissions/" },
    { "id": "starring", "text": "Starring", "note": "List and manage repositories a user is starring.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-starring" },
    { "id": "watching", "text": "Watching ", "note": "List and change repositories a user is subscribed to.", "accessLevel": ["No access", "Read-only", "Read & Write"], "link": "https://docs.github.com/v3/apps/permissions/#permission-on-watching" }
];
interface IUserPermissionsProps {
    setTemplateState: any;
    onBoardingTemplateObject: onBoardingObject;
}
class UserPermissions extends React.Component<IUserPermissionsProps> {
    constructor(props: any) {
        super(props);
        this.initializeUserPermissions();
    }

    initializeUserPermissions() {
        let onBoardingTemplateObject = this.props.onBoardingTemplateObject;
        userPermissions.map(userPermission => {
            onBoardingTemplateObject.userPermissions[userPermission.id] = userPermission.accessLevel[0]
        });
        this.props.setTemplateState(onBoardingTemplateObject);
    }

    render() {
        let userPerm: any = [];
        let accessLevel: any = [];
        userPermissions.map(userPermission => {
            accessLevel = [];
            userPermission.accessLevel.map((access, index) => {
                let hidden = true;
                if (access === this.props.onBoardingTemplateObject.userPermissions[userPermission.id]) {
                    hidden = false;
                }
                accessLevel.push(
                    <SelectMenu.Item key={userPermission.text + " " + access} id={userPermission.id + access + index} href="#" onClick={() => this.onAccessLevelClick(userPermission.id, access)}>
                        <div className="input-access">
                            <div className="input-access-tick">
                                <img aria-label="checked" hidden={hidden} className="tick-img" alt="" src={tick} width="16px" height="16px" />
                            </div>
                            <div aria-checked={true} className="input-access-text">
                                {access}
                            </div>
                        </div>
                    </SelectMenu.Item>
                );
            });
            userPerm.push(
                <div key={userPermission.id} className="user-permission-item tableobject">
                    <div className="tableobject-cell-primary">
                        <div className="text-bold">{userPermission.text}
                            <a className="link" aria-label="info" href={userPermission.link}>
                                <img className="info-img" alt="" src={icon} width="16px" height="16px" />
                            </a>
                        </div>
                        <p className="note">{userPermission.note}</p>
                    </div>
                    <div className="tableobject-cell-intermediate">
                        Access:
                    </div>
                    <div className="tableobject-cell-secondary">
                        <SelectMenu>
                            <Button as="summary" className="selectmenu-button">
                                <span className="selectmenu-button-text">{this.props.onBoardingTemplateObject.userPermissions[userPermission.id]}</span>
                                <img className="down-img" alt="" src={down} />
                            </Button>
                            <SelectMenu.Modal width="212px">
                                <SelectMenu.Header>
                                    <span>{Constants.SelectMenuHeading}</span>
                                    <this.AccessLevelCloseButton />
                                </SelectMenu.Header>
                                <SelectMenu.List>
                                    {accessLevel}
                                </SelectMenu.List>
                            </SelectMenu.Modal>
                        </SelectMenu>
                    </div>
                </div>
            )
        });
        return (
            <div className="permissions">
                <div className="user-permissions" aria-label="user-permissions">
                    <div className="user-permissions-description">
                        <p>
                            {Constants.UserPermissionDescription1}
                            <a href={Constants.UserPermissionDescriptionLink}>{Constants.UserPermissionDescriptionForLink}</a>
                            {Constants.UserPermissionDescription2}
                        </p>
                    </div>
                    {userPerm}
                </div>
            </div>

        )
    }

    onAccessLevelClick(id: string, accessValue: string) {
        let onBoardingTemplateObject = this.props.onBoardingTemplateObject;
        onBoardingTemplateObject.userPermissions[id] = accessValue;
        this.props.setTemplateState(onBoardingTemplateObject);
    }

    AccessLevelCloseButton = () => {
        const menuContext = React.useContext(SelectMenu.MenuContext);
        return (
            <button type="button" aria-label="close menu" className="cross-img-button" onClick={() => {
                menuContext.setOpen(false);
            }}>
                <img className="cross-img" alt="" src={cross} width="10px" height="10px" />
            </button>
        );
    }
}

export default UserPermissions
