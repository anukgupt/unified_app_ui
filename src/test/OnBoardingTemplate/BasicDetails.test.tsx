import React from 'react';
import renderer from 'react-test-renderer';
import BasicDetails from '../../components/onBoardingTemplate/BasicDetails';
import { onBoardingObject } from "../../components/onBoardingTemplate/OnBoardingTemplate";
import OnBoardingTemplate from '../../components/onBoardingTemplate/OnBoardingTemplate';
import { shallow, mount } from "enzyme";
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });
const onBoardingTemplateObject: onBoardingObject = {
    clientId: "",
    clientIdError: false,
    webhookURL: "",
    webhookURLError: false,
    callbackURL: "",
    callbackURLError: false,
    repoPermissions: {},
    orgPermissions: {},
    userPermissions: {},
    showOnBoardingTemplate: true,
    isUserAuthorized: false,
    isTemplateSucessfullySaved: false,
    selectedTabIndex: 0,
    isUserAdminFlowInProcess: true
}

const onBoardingUI = mount(<OnBoardingTemplate />);
const onClientIdChange = jest.fn(() => {
    onBoardingUI.setState({
        clientId: "test clientid",
    });
});
const onWebhookURLChange = jest.fn(() => {
    onBoardingUI.setState({
        webhookURL: "test webhookurl",
    });
});
const onCallbackURLChange = jest.fn(() => {
    onBoardingUI.setState({
        callbackURL: "test callbackurl",
    });
});

test('renders correctly all components with no errors', () => {
    const basicDetails = renderer.create(<BasicDetails
        onBoardingTemplateObject={onBoardingTemplateObject} onClientIdChange={onClientIdChange} onWebhookURLChange={onWebhookURLChange} onCallbackURLChange={onCallbackURLChange} />).toJSON();
    expect(basicDetails).toMatchSnapshot()
});

test('render input clientid', () => {
    const basicDetails = mount(<BasicDetails
        onBoardingTemplateObject={onBoardingTemplateObject} onClientIdChange={onClientIdChange} onWebhookURLChange={onWebhookURLChange} onCallbackURLChange={onCallbackURLChange} />);
    basicDetails.find('input').at(0).simulate('change', 'test clientid');
    expect(onClientIdChange).toHaveBeenCalledTimes(1);
    expect(onBoardingUI.state()).toStrictEqual({ "clientId": "test clientid", "currentAuthority": "", "error": null, "isAuthenticated": false, "user": {} });
});

test('render input webhookurl', () => {
    let basicDetails = mount(<BasicDetails
        onBoardingTemplateObject={onBoardingTemplateObject} onClientIdChange={onClientIdChange} onWebhookURLChange={onWebhookURLChange} onCallbackURLChange={onCallbackURLChange} />);
    basicDetails.find('input').at(1).simulate('change', 'test webhookurl');
    expect(onWebhookURLChange).toHaveBeenCalledTimes(1);
    expect(onBoardingUI.state()).toStrictEqual({ "clientId": "test clientid", "webhookURL": "test webhookurl", "currentAuthority": "", "error": null, "isAuthenticated": false, "user": {} });
});

test('render input callbackurl', () => {
    const basicDetails = mount(<BasicDetails
        onBoardingTemplateObject={onBoardingTemplateObject} onClientIdChange={onClientIdChange} onWebhookURLChange={onWebhookURLChange} onCallbackURLChange={onCallbackURLChange} />);
    basicDetails.find('input').at(2).simulate('change', 'test callbackurl');
    expect(onCallbackURLChange).toHaveBeenCalledTimes(1);
    expect(onBoardingUI.state()).toStrictEqual({ "clientId": "test clientid", "webhookURL": "test webhookurl", "callbackURL": "test callbackurl", "currentAuthority": "", "error": null, "isAuthenticated": false, "user": {} });
});

test('render buttons', () => {
    expect(onBoardingUI.find('.create-configuration')).toBeTruthy();
    expect(onBoardingUI.find('.previous-configuration')).toBeTruthy();
    expect(onBoardingUI.find('.next-configuration')).toBeTruthy();
})
