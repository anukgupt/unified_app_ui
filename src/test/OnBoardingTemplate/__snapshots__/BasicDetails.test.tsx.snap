// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly all components with no errors 1`] = `
<div
  className="basic-details"
>
  <div
    className="client-id"
  >
    <div
      className=""
    >
      <div
        className="subHead"
      >
        Client Id
      </div>
    </div>
    <dl>
      <dt>
        <label
          className="input-label"
          htmlFor="clientId"
        >
          Client Id
        </label>
      </dt>
      <dd>
        <input
          className="input-text"
          id="clientId"
          onChange={[MockFunction]}
          type="text"
          value=""
        />
        <p
          className="note"
        >
          ClientId for onboarding registration
        </p>
      </dd>
    </dl>
  </div>
  <div
    className="webhook-url"
  >
    <div
      className=""
    >
      <div
        className="subHead"
      >
        Webhook
      </div>
    </div>
    <dl>
      <dt>
        <label
          className="input-label"
          htmlFor="webhookURL"
        >
          WebhookURL
        </label>
      </dt>
      <dd>
        <input
          className="input-text"
          id="webhookURL"
          onChange={[MockFunction]}
          type="text"
          value=""
        />
        <p
          className="note"
        >
          Events will POST to this URL. Read our 
          <a
            href="https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/about-webhooks"
            target="_blank"
          >
            webhook documentation
          </a>
           for more information.
        </p>
      </dd>
    </dl>
  </div>
  <div
    className="callback-url"
  >
    <div
      className=""
    >
      <div
        className="subHead"
      >
        Callback
      </div>
    </div>
    <dl>
      <dt>
        <label
          className="input-label"
          htmlFor="callbackURL"
        >
          CallbackURL
        </label>
      </dt>
      <dd>
        <input
          className="input-text"
          id="callbackURL"
          onChange={[MockFunction]}
          type="text"
          value=""
        />
        <p
          className="note"
        >
          The full URL to redirect to after a user authorizes an installation.
        </p>
      </dd>
    </dl>
  </div>
</div>
`;
